### ISSUE TITLE
Decimal methods not supported

### ISSUE THREAD

**Issue by grahampcharles** [prisma/prisma#14506]
### Bug description

According to the docs, Decimal comes from the library [Decimal.js](https://mikemcl.github.io/decimal.js/). However, the methods from Decimal.js produce an error.

```
export const FormatPrice = (input: Decimal | null) => {
    const valueString = input === null ? "--" : input.toFixed(2);
    return valueString;
};
```

This produces an unhandled runtime error: `input.toFixed is not a function`.

Is prisma's Decimal a complete implementation of [Decimal.js](https://mikemcl.github.io/decimal.js/) or some sort of subset?

### How to reproduce

In a project that has a Decimal data type, create a function to format Decimal values:
```
export const FormatPrice = (input: Decimal | null) => {
    const valueString = input === null ? "--" : input.toFixed(2);
    return valueString;
};
```
The function will not compile.

### Environment & setup

- OS: Ubuntu, Windows
- Database: MySQL
- Node.js version: 16.13.0

### Prisma Version

```
prisma                  : 4.0.0
@prisma/client          : 4.0.0
Current platform        : windows
Query Engine (Node-API) : libquery-engine da41d2bb3406da22087b849f0e911199ba4fbf11 (at node_modules\.pnpm\@prisma+engines@3.16.0-49.da41d2bb3406da22087b849f0e911199ba4fbf11\node_modules\@prisma\engines\query_engine-windows.dll.node)
Migration Engine        : migration-engine-cli da41d2bb3406da22087b849f0e911199ba4fbf11 (at node_modules\.pnpm\@prisma+engines@3.16.0-49.da41d2bb3406da22087b849f0e911199ba4fbf11\node_modules\@prisma\engines\migration-engine-windows.exe)
Introspection Engine    : introspection-core da41d2bb3406da22087b849f0e911199ba4fbf11 (at node_modules\.pnpm\@prisma+engines@3.16.0-49.da41d2bb3406da22087b849f0e911199ba4fbf11\node_modules\@prisma\engines\introspection-engine-windows.exe)
Format Binary           : prisma-fmt da41d2bb3406da22087b849f0e911199ba4fbf11 (at node_modules\.pnpm\@prisma+engines@3.16.0-49.da41d2bb3406da22087b849f0e911199ba4fbf11\node_modules\@prisma\engines\prisma-fmt-windows.exe)
Default Engines Hash    : da41d2bb3406da22087b849f0e911199ba4fbf11
Studio                  : 0.465.0
Preview Features        : referentialIntegrity
```

**Comment by garrensmith** [comment]
@grahampcharles could you provide us with your schema and the query you are doing to get this error?

**Comment by grahampcharles** [comment]
Sure, here's the relevant part of the schema:

```
generator client {
  provider        = "prisma-client-js"
  previewFeatures = ["referentialIntegrity"]
}

datasource db {
  provider             = "mysql"
  url                  = env("DATA_DATABASE_URL")
  referentialIntegrity = "prisma"
}

model Item {
  id             Int       @id @default(autoincrement())
  Variety        String?
  Retail         Decimal?  @db.Decimal(10, 2)
}
```

The query from `getServerSideProps`:

```
export const getServerSideProps: GetServerSideProps = async (context) => {

    let item = await prisma.item.findFirst({
    where: { id: Number(id) }}) as Item | null;

     return {
        props:
        {
            item: JSON.parse(JSON.stringify(item))
        }
    }
}
```

Do you think the problem is the parse/stringify step? I've gotten into the habit of doing that since prisma seems to fail on serializing most types (like dates, for example). But maybe I'm overapplying it?

**Comment by ChristianIvicevic** [comment]
Even without gSSP I have the same problems with `Decimal`s not having the instance methods. My current workaround is to use the static methods such as `Prisma.Decimal.floor(valueFromDbViaPrisma)` or even "worse" `new Prisma.Decimal(resources.metal).toNumber()`.

**Comment by grahampcharles** [comment]
Yes, only the `new` version seems to work -- `Prisma.Decimal` objects have no instance methods. (If I try `Prisma.Decimal.toNumber(value)` I get a TypeScript error (`Property 'toNumber' does not exist on type 'typeof Decimal'.ts(2339)`). I've been using these utility functions:

```
export function DecimalToNumber(
    value: Prisma.Decimal | null
): number | undefined {
    return value === null ? undefined : new Prisma.Decimal(value).toNumber();
}

export function NumberToDecimal(
    value: number | undefined
): Prisma.Decimal | null {
    return value === undefined ? null : new Prisma.Decimal(value);
}

```

**Comment by SevInf** [comment]
@ChristianIvicevic @grahampcharles I don't think this problem has anything to do with prisma. `JSON.parse(JSON.stringify())` loop is where decimal instance gets lost. `Decimal` is not a thing in `JSON`, when you do `stringify` it is converted to string. `JSON.parse` then does not know anything about `Decimal` as well and parses value as string. You'll have to manually convert decimal values to decimal instances after `JSON.parse`, there is no way around it. 

Since it does not seem to be Prisma issue, I'll close it for now. Feel free to reopen if you manage to reproduce it without `JSON.parse(JSON.stringify())` part.

**Comment by ChristianIvicevic** [comment]
Thanks for the clarification, it hasn't been clear that this is "intended" due to that limitation. However it is somewhat disingenuous that the typings claim that the return type of a query is a `Decimal` when in fact it is not. Have you possibly considered evaluating something like `superjson` in order to ensure a `Decimal` actually stays a `Decimal`?

I'll take a look at proposing an update to docs to document that behavior, won't promise anything though.

**Comment by ChristianIvicevic** [comment]
Please excuse the double post, but I just checked the readme of [`superjson`](https://github.com/blitz-js/superjson) and it shows how to include a babel plugin to enable improved (de)serialization for gSSP. I am just unsure whether this would disable swc though. Just wanted to let anybody know that this is an option who's running into this issue.

**Comment by grahampcharles** [comment]
Yeah, I get that I was posting about the wrong issue, but I still think this is an issue. Prisma seems to depend on types (like Decimal) that cannot be serialized... that's a serious usability issue and requires all sorts of hoop-jumping.

**Comment by SevInf** [comment]
Well, no types except `string`, `number`, `boolean` and `null` are JSON-serializable and de-serializable without extra steps. So, the only way to make it work out of the box will be returning only those primitive types from prisma queries, which would be an issue for many other users and a huge breaking change.

As an alternative, maybe prisma iteself could provide serialization and de-serialization routines, that will take care of those extra types. Something along the lines of:

```ts
import { toJson, fromJson } from '@prisma/client`

export const getServerSideProps: GetServerSideProps = async (context) => {
    let item = await prisma.item.findFirst({ where: { id }});

     return {
        props:
        {
            item: toJson(item)
        }
    }
}
```

Will something like this be helpful? If so, we can have a separate feature request for this.

**Comment by grahampcharles** [comment]
Hm. Well first, there's this: Today I learned that in Next.js, prisma objects must cross the getSSP -> component divide. (A similar thing would happen in vanilla React, I think?) It's not readily clear that that step is going to involve serialization. So first off, I'd say that needs to be documented.

Second, it would seem that nearly everyone who uses prisma with React is going to run into a JSON serialization problem unless they're only using primitive data types -- and since prisma makes use of `Decimal` instead of `number`, that includes anyone using any kind of floating-point number. What does everyone else do in this situation? I ran into the problem on a date field and found that The Internet said to just put a `parse(stringify())` around the object, but clearly that's a bad idea. 

Anyway, this is clearly a quirk of interaction between prisma (among other packages!) and React. Given that, yes, I think a helper function to fix the problem (an adapted superjson?) would be lovely. At least, some documentation of the issue would be a really good start.
